libs/popupmanager.py

```diff
@@ -0,0 +1,209 @@
+        arg_span = self.signature_help["applicableSpan"]
+        span_start = view.text_point(
+                                    arg_span["start"]["line"] - 1,
+                                    arg_span["start"]["offset"] - 2)
+        span_end = view.text_point(
+                                    arg_span["end"]["line"] - 1,
+                                    arg_span["end"]["offset"] - 1)
+        arg_region = sublime.Region(span_start, span_end)
+        view.add_regions('argSpan', [arg_region],
+                         flags=sublime.HIDDEN)
+        # To view region, set to: scope='comments', flags=sublime.DRAW_EMPTY)
+
+        self.display()
+
+    def display(self):
+        popup_parts = self.get_current_signature_template()
+        popup_text = PopupManager.html_template.substitute(popup_parts)
+
+        log.debug('Displaying signature popup')
+        if not self.current_view.is_popup_visible():
+            self.current_view.show_popup(
+                popup_text,
+                sublime.COOPERATE_WITH_AUTO_COMPLETE,
+                on_navigate=self.on_navigate,
+                on_hide=self.on_hidden,
+                max_width=800)
+        else:
+            self.current_view.update_popup(popup_text)
+
+    def move_next(self):
+        if not self.signature_help:
+            return
+        self.signature_index += 1
+        if self.signature_index >= len(self.signature_help["items"]):
+            self.signature_index = len(self.signature_help["items"]) - 1
+        self.display()
+
+    def move_prev(self):
+        if not self.signature_help:
+            return
+        self.signature_index -= 1
+        if self.signature_index < 0:
+            self.signature_index = 0
+        self.display()
+
+    def on_navigate(self, loc):
+        # Clicked the overloads link.  Dismiss this popup and show the panel
+        view = self.current_view
+        self.on_close_popup()
+        view.run_command('typescript_signature_panel')
+
+    def on_hidden(self):
+        log.debug('In popup on_hidden handler')
+        if not self.current_view:
+            return
+
+        cursor_region = self.current_view.sel()[0]
+        arg_regions = self.current_view.get_regions('argSpan')
+        if len(arg_regions):
+            argSpan = self.current_view.get_regions('argSpan')[0]
+            if argSpan.contains(cursor_region):
+                log.debug('Was hidden while in region.  Redisplaying')
+                # Occurs on left/right movement.  Rerun to redisplay popup.
+                self.display()
+        else:
+            # Cleanup
+            self.on_close_popup()
+
+    def on_close_popup(self):
+        """ Call whenever the view loses focus, of the region is exited """
+        if self.current_view:
+            self.current_view.erase_regions('argSpan')
+            self.current_view.hide_popup()
+            self.current_view = None
+        self.signature_help = None
+
+    def is_active(self):

```

Comment:
Comment this.  Just means a session is currently active.  current_view is set iff this is true.